headers.go

```diff
@@ -0,0 +1,175 @@
+package cose
+
+import (
+	"fmt"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+// COSE Header labels registered in the IANA "COSE Header Parameters" registry.
+//
+// Reference: https://www.iana.org/assignments/cose/cose.xhtml#header-parameters
+const (
+	HeaderLabelAlgorithm         int64 = 1
+	HeaderLabelCritical          int64 = 2
+	HeaderLabelContentType       int64 = 3
+	HeaderLabelKeyID             int64 = 4
+	HeaderLabelCounterSignature  int64 = 7
+	HeaderLabelCounterSignature0 int64 = 9
+	HeaderLabelX5Bag             int64 = 32
+	HeaderLabelX5Chain           int64 = 33
+	HeaderLabelX5T               int64 = 34
+	HeaderLabelX5U               int64 = 35
+)
+
+// ProtectedHeader contains parameters that are to be cryptographically
+// protected.
+type ProtectedHeader map[interface{}]interface{}
+
+// MarshalCBOR encodes the protected header into a CBOR bstr object.
+// A zero-length header is encoded as a zero-length string rather than as a
+// zero-length map (encoded as h'a0').
+func (h ProtectedHeader) MarshalCBOR() ([]byte, error) {
+	var encoded []byte
+	if len(h) == 0 {
+		encoded = []byte{}
+	} else {
+		var err error
+		encoded, err = encMode.Marshal(map[interface{}]interface{}(h))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return encMode.Marshal(encoded)
+}
+
+// UnmarshalCBOR decodes a CBOR bstr object into ProtectedHeader.
+func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error {
```

Comment:
This function should check if the `alg` header is present and, if so, cast its value to an `Algorithm`.

Reply:
That's doable.